File: src/Cdn/Certificates/Provider/Cloudflare.php
<?php

namespace Utopia\Cdn\Certificates\Provider;

use Psr\Http\Client\ClientExceptionInterface;
use Psr\Http\Client\ClientInterface;
use Utopia\Cdn\Certificates\Provider;
use Utopia\Cdn\Domain;
use Utopia\Cdn\Exception\UnsupportedOperation;
use Utopia\Client;
use Utopia\Client\Adapter\Curl\Client as CurlAdapter;
use Utopia\Psr7\ContentType;
use Utopia\Psr7\Header;
use Utopia\Psr7\Method;
use Utopia\Psr7\Request\Factory as RequestFactory;

class Cloudflare implements Provider
{
    private readonly ClientInterface $client;

    public function __construct(
        private readonly string $zoneId,
        private readonly string $apiToken,
        ?ClientInterface $client = null,
        private readonly string $apiBase = 'https://api.cloudflare.com/client/v4',
    ) {
        $this->client = $client ?? new Client(new CurlAdapter());
    }

    public function issueCertificate(string $certName, string $domain, ?string $domainType): ?string
    {
        $domain = Domain::validate($domain);
        $result = $this->request(Method::POST, $this->hostnamesPath(), [
            'hostname' => $domain,
            'ssl' => [
                'method' => 'http',
                'type' => 'dv',
                'wildcard' => false,
            ],
        ]);

        if ($this->isDuplicate($result)) {
            return null;
        }

        $this->assertSuccess('create Cloudflare custom hostname', $result, [201]);

        return null;
    }

    public function isInstantGeneration(string $domain, ?string $domainType): bool
    {
        Domain::validate($domain);

        return true;
    }

    public function getCertificateStatus(string $domain, ?string $domainType): string
    {
        throw new UnsupportedOperation('Certificate status retrieval is not supported by the Cloudflare provider.');
    }

    public function isRenewRequired(string $domain, ?string $domainType): bool
    {
        return $this->findHostname(Domain::validate($domain)) === null;
    }

    public function deleteCertificate(string $domain, ?string $domainType = null): void
    {
        $hostname = $this->findHostname(Domain::validate($domain));
        if ($hostname === null) {
            return;
        }

        $id = $hostname['id'] ?? null;
        if (!\is_string($id) || $id === '') {
            throw new \RuntimeException('Cloudflare custom hostname response was missing an ID.');
        }

        $result = $this->request(Method::DELETE, $this->hostnamesPath() . '/' . rawurlencode($id));
        $this->assertSuccess('delete Cloudflare custom hostname', $result);
    }

    /** @return array<string, mixed>|null */
    private function findHostname(string $domain): ?array
    {
        $result = $this->request(Method::GET, $this->hostnamesPath() . '?' . http_build_query(['hostname' => $domain]));
        $this->assertSuccess('fetch Cloudflare custom hostnames', $result);

        if (!\is_array($result['response'])) {
            throw new \RuntimeException('Cloudflare custom hostname response was not valid JSON.');
        }

        $hostnames = $result['response']['result'] ?? null;
        if (!\is_array($hostnames)) {
            throw new \RuntimeException('Cloudflare custom hostname response was missing its result list.');
        }

        foreach ($hostnames as $hostname) {
            if (\is_array($hostname) && ($hostname['hostname'] ?? null) === $domain) {
                return $hostname;
            }
        }

        return null;
    }

    private function hostnamesPath(): string
    {
        return '/zones/' . $this->zoneId . '/custom_hostnames';
    }

    /** @param array{statusCode:int,response:array<string, mixed>|string|null,error:string|null} $result */
    private function isDuplicate(array $result): bool
    {
        return \is_array($result['response']) && (($result['response']['errors'][0]['code'] ?? null) === 1406);
    }

    /**
     * @param array{statusCode:int,response:array<string, mixed>|string|null,error:string|null} $result
     * @param array<int, int>|null $expectedStatuses
     */
    private function assertSuccess(string $operation, array $result, ?array $expectedStatuses = null): void
    {
        $httpSuccess = $expectedStatuses === null
            ? $result['statusCode'] >= 200 && $result['statusCode'] < 300
            : \in_array($result['statusCode'], $expectedStatuses, true);
        $envelopeSuccess = !\is_array($result['response']) || !\array_key_exists('success', $result['response']) || $result['response']['success'] === true;

        if (!$httpSuccess || !$envelopeSuccess) {
            $message = $result['error'];
            if (\is_array($result['response'])) {
                $message ??= $result['response']['errors'][0]['message'] ?? null;
            }

            throw new \RuntimeException('Failed to ' . $operation . ' with status ' . $result['statusCode'] . ': ' . ($message ?? 'Unknown Cloudflare error'));
        }
    }

    /**
     * @param array<string, mixed>|null $body
     * @return array{statusCode:int,response:array<string, mixed>|string|null,error:string|null}
     */
    private function request(string $method, string $path, ?array $body = null): array
    {
        $factory = new RequestFactory();
        $request = $body === null
            ? $factory->createRequest($method, $this->apiBase . $path)
            : $factory->json($method, $this->apiBase . $path, $body);
        $request = $request
            ->withHeader(Header::USER_AGENT, 'Utopia CDN Cloudflare Certificates Provider')
            ->withHeader(Header::AUTHORIZATION, 'Bearer ' . $this->apiToken)
            ->withHeader(Header::CONTENT_TYPE, ContentType::JSON);

        try {
            $response = $this->client->sendRequest($request);
        } catch (ClientExceptionInterface $error) {
            return ['statusCode' => 0, 'response' => null, 'error' => $error->getMessage()];
        }

        $contents = (string) $response->getBody();

        try {
            $decoded = json_decode($contents, true, flags: JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            $decoded = $contents;
        }

        return ['statusCode' => $response->getStatusCode(), 'response' => $decoded, 'error' => null];
    }
}

File: 3.x.x/crates-utopia/cdn/src/certificates/provider/cloudflare.rs
//! PHP `Utopia\Cdn\Certificates\Provider\Cloudflare`.

use std::sync::Arc;

use serde_json::{json, Value};

use super::Provider;
use crate::http::{
    default_client, php_http_build_query, php_rawurlencode, send, HttpClient, RequestResult,
};
use crate::{CdnError, Domain, UnsupportedOperation};

/// Cloudflare for `SaaS` custom-hostname certificates.
pub struct Cloudflare {
    zone_id: String,
    api_token: String,
    api_base: String,
    client: Arc<dyn HttpClient>,
}

impl std::fmt::Debug for Cloudflare {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Cloudflare")
            .field("zone_id", &self.zone_id)
            .field("api_base", &self.api_base)
            .finish_non_exhaustive()
    }
}

impl Cloudflare {
    #[must_use]
    pub fn new(zone_id: impl Into<String>, api_token: impl Into<String>) -> Self {
        Self {
            zone_id: zone_id.into(),
            api_token: api_token.into(),
            api_base: "https://api.cloudflare.com/client/v4".into(),
            client: Arc::new(default_client()),
        }
    }

    #[must_use]
    pub fn with_client(mut self, client: Arc<dyn HttpClient>) -> Self {
        self.client = client;
        self
    }

    #[must_use]
    pub fn with_api_base(mut self, api_base: impl Into<String>) -> Self {
        self.api_base = api_base.into();
        self
    }

    fn hostnames_path(&self) -> String {
        format!("/zones/{}/custom_hostnames", self.zone_id)
    }

    fn request(&self, method: &str, path: &str, body: Option<&Value>) -> RequestResult {
        let headers = [
            (
                "User-Agent",
                "Utopia CDN Cloudflare Certificates Provider".to_owned(),
            ),
            ("Authorization", format!("Bearer {}", self.api_token)),
            ("Content-Type", "application/json".to_owned()),
        ];
        send(
            self.client.as_ref(),
            method,
            &format!("{}{path}", self.api_base),
            &headers,
            body,
        )
    }

    fn is_duplicate(result: &RequestResult) -> bool {
        result
            .response
            .get("errors")
            .and_then(Value::as_array)
            .and_then(|errors| errors.first())
            .and_then(|error| error.get("code"))
            .and_then(Value::as_i64)
            == Some(1406)
    }

    fn assert_success(
        operation: &str,
        result: &RequestResult,
        expected: Option<&[u16]>,
    ) -> Result<(), CdnError> {
        let http_ok = match expected {
            Some(list) => list.contains(&result.status),
            None => (200..300).contains(&result.status),
        };
        let envelope_ok = result.response.as_object().map_or(true, |object| {
            !object.contains_key("success") || object.get("success") == Some(&Value::Bool(true))
        });
        if !http_ok || !envelope_ok {
            let message = result
                .error
                .clone()
                .or_else(|| {
                    result
                        .response
                        .get("errors")
                        .and_then(Value::as_array)
                        .and_then(|errors| errors.first())
                        .and_then(|error| error.get("message"))
                        .and_then(Value::as_str)
                        .map(str::to_owned)
                })
                .unwrap_or_else(|| "Unknown Cloudflare error".into());
            return Err(CdnError::runtime(format!(
                "Failed to {operation} with status {}: {message}",
                result.status
            )));
        }
        Ok(())
    }

    fn find_hostname(&self, domain: &str) -> Result<Option<Value>, CdnError> {
        let query = php_http_build_query(&[("hostname", domain)]);
        let result = self.request("GET", &format!("{}?{query}", self.hostnames_path()), None);
        Self::assert_success("fetch Cloudflare custom hostnames", &result, None)?;
        if !result.response.is_object() {
            return Err(CdnError::runtime(
                "Cloudflare custom hostname response was not valid JSON.",
            ));
        }
        let Some(hostnames) = result.response.get("result").and_then(Value::as_array) else {
            return Err(CdnError::runtime(
                "Cloudflare custom hostname response was missing its result list.",
            ));
        };
        for hostname in hostnames {
            if hostname.get("hostname").and_then(Value::as_str) == Some(domain) {
                return Ok(Some(hostname.clone()));
            }
        }
        Ok(None)
    }
}

impl Provider for Cloudflare {
    fn issue_certificate(
        &self,
        _cert_name: &str,
        domain: &str,
        _domain_type: Option<&str>,
    ) -> Result<Option<String>, CdnError> {
        let domain = Domain::validate(domain)?;
        let result = self.request(
            "POST",
            &self.hostnames_path(),
            Some(&json!({
                "hostname": domain,
                "ssl": {"method": "http", "type": "dv", "wildcard": false},
            })),
        );
        if Self::is_duplicate(&result) {
            return Ok(None);
        }
        Self::assert_success("create Cloudflare custom hostname", &result, Some(&[201]))?;
        Ok(None)
    }

    fn is_instant_generation(
        &self,
        domain: &str,
        _domain_type: Option<&str>,
    ) -> Result<bool, CdnError> {
        Domain::validate(domain)?;
        Ok(true)
    }

    fn get_certificate_status(
        &self,
        _domain: &str,
        _domain_type: Option<&str>,
    ) -> Result<String, CdnError> {
        Err(UnsupportedOperation(
            "Certificate status retrieval is not supported by the Cloudflare provider.".into(),
        )
        .into())
    }

    fn is_renew_required(
        &self,
        domain: &str,
        _domain_type: Option<&str>,
    ) -> Result<bool, CdnError> {
        Ok(self.find_hostname(&Domain::validate(domain)?)?.is_none())
    }

    fn delete_certificate(&self, domain: &str, _domain_type: Option<&str>) -> Result<(), CdnError> {
        let Some(hostname) = self.find_hostname(&Domain::validate(domain)?)? else {
            return Ok(());
        };
        let id = hostname.get("id").and_then(Value::as_str).unwrap_or("");
        if id.is_empty() {
            return Err(CdnError::runtime(
                "Cloudflare custom hostname response was missing an ID.",
            ));
        }
        let encoded = php_rawurlencode(id);
        let result = self.request(
            "DELETE",
            &format!("{}/{encoded}", self.hostnames_path()),
            None,
        );
        Self::assert_success("delete Cloudflare custom hostname", &result, None)
    }
}
